# Data Processing Pipeline

**Author:** Team 5  
**Course:** Advanced Data Science  
**Date:** Fall 2025

---

## Executive Summary

This notebook processes raw gold and economic data to create a clean dataset for modeling.

**Research Question:** Should investors allocate capital to gold or stocks?

**Prediction Target:** Will gold outperform the S&P 500 over the next 90 days?

---

## Data Sources

### Daily Market Data (Investing.com)
- **Gold Futures:** Daily gold futures prices
- **S&P 500 Index:** US large-cap stock market performance
- **Silver Futures:** Alternative precious metal indicator
- **USD Index (DXY):** US dollar strength vs basket of currencies

### Macroeconomic Indicators (FRED - Federal Reserve Economic Data)
- **VIX:** Market volatility/fear index
- **NASDAQ:** Technology-heavy stock index
- **CPI:** Consumer Price Index (inflation)
- **Federal Funds Rate:** Central bank policy rate
- **M2 Money Supply:** Broad measure of money in circulation
- **Unemployment Rate:** Labor market health
- **Crude Oil:** Energy commodity prices
- **US Treasury Yields:** Government bond interest rates

### Rationale for Data Selection

Gold and stocks respond differently to economic conditions:
- **Gold:** Safe haven during crises, inflation hedge
- **Stocks:** Perform well during economic expansions
- **Macro factors:** Drive asset allocation decisions (inflation, rates, volatility)

**Mixed frequency approach:** Daily market data captures price dynamics, monthly macro data captures structural economic trends.

---

## Key Improvements from Initial Submission

1. **Expanded dataset:** 220 → ~4,548 observations 
2. **More features:** 11 datasets merged
3. **Modular pipeline:** Reproducible data processing via `data_pipeline.py`
4. **Professional structure:** Follows data science lifecycle best practices

In [1]:
# SETUP: Import Libraries and Configure Environment

import pandas as pd
import numpy as np
import sys
from pathlib import Path

# Add src directory to Python path for module imports
sys.path.append('../src')

# Import custom pipeline module
from data_pipeline import build_features, save_processed_data

print("Libraries loaded successfully")
print(f"Python version: {sys.version.split()[0]}")

Libraries loaded successfully
Python version: 3.11.13


## Step 1: Execute Data Pipeline

The build_features() function orchestrates the complete data processing workflow:

### Pipeline Stages

1. **Load raw datasets:** Reads 11 CSV files from `data/raw/`
2. **Clean data:** Removes commas from numbers, converts percentages to floats, drops OHLC/volume
3. **Merge daily data:** Inner join on core assets (gold, S&P, silver), left join on supplementary data
4. **Forward-fill monthly data:** Propagates monthly macro indicators to daily frequency
5. **Engineer features:** Creates returns, moving averages, volatility metrics, derived indicators
6. **Create target:** Binary indicator (1 = gold outperforms S&P 500 over 90 days)
7. **Filter and clean:** Removes rows without valid target

### Time Window

**2006-2024** (limited by data availability)

### Output

Clean dataset with:
- ~4,548 daily observations
- 30+ engineered features
- Binary target for classification

In [7]:
# EXECUTE PIPELINE: Process All Raw Data

df = build_features(
    data_dir='../data/raw',
    start_date='2006-01-03',  # First available gold data
    end_date='2024-04-11',     # Last available S&P 500 data
    target_horizon=90          # 90-day forward prediction window
)

print("\nPIPELINE OUTPUT SUMMARY")
print(f"Dataset shape: {df.shape[0]:,} rows × {df.shape[1]} columns")
print(f"Date range: {df['date'].min().date()} to {df['date'].max().date()}")
print(f"Total trading days: {len(df):,}")

Loading daily market data...
  Gold: 5000 rows
  S&P 500: 5000 rows
  NASDAQ: 14274 rows
  VIX: 5193 rows

Column names after standardization:
  Treasury: ['treasury_dgs1']
  NASDAQ: ['nasdaq_nasdaqcom']
  VIX: ['vix_vixcls']

Merging daily datasets...
  Merging core assets (gold, S&P, silver)...
    After core merge: 4548 rows
  Merging supplementary data...
  After merge: 4548 rows
  After date filter (2006-01-03 to 2024-04-11): 4548 rows

Engineering features...
  After feature engineering: 33 columns

Loading monthly macro data...
Merging monthly data...
Creating derived macro features...

Creating target variable...
PIPELINE COMPLETE
Final dataset: 4548 rows, 42 columns
Date range: 2006-01-03 to 2024-04-11
Target distribution: 42.8% positive

Features with missing values:
  oil_return_60d: 60 (1.3%)
  silver_return_60d: 60 (1.3%)
  sp500_vol_60d: 60 (1.3%)
  gold_vol_60d: 60 (1.3%)
  oil_vol_60d: 60 (1.3%)

PIPELINE OUTPUT SUMMARY
Dataset shape: 4,548 rows × 42 columns
Date range:

## Step 2: Save Processed Data

Export cleaned dataset for use in EDA (02_eda.ipynb) and modeling (03_modeling.ipynb) notebooks.

In [11]:
# SAVE: Export Processed Dataset

save_processed_data(df, output_dir='../data/processed', filename='gold_features.csv')

print(f"\nTotal features: {df.shape[1] - 2} (excluding date and target)")


Saved to: ../data/processed/gold_features.csv

Total features: 40 (excluding date and target)


## Step 3: Data Quality Assessment

Validate target distribution, check for missing values, and assess class balance.

In [12]:
# QUALITY CHECK: Target Distribution and Missing Values

print("DATA QUALITY SUMMARY")

# Target variable distribution
print("\nTarget Variable Distribution:")
target_positive = df['target'].sum()
target_negative = (1 - df['target']).sum()
target_pct = df['target'].mean() * 100

print(f"- Gold outperforms (target=1): {target_positive:,} ({target_pct:.1f}%)")
print(f"- S&P outperforms (target=0): {target_negative:,} ({(100-target_pct):.1f}%)")

if 40 <= target_pct <= 60:
    print(f"Class balance: Good (within 40-60% range)")
else:
    print(f"Class balance: Acceptable (slight imbalance)")

# Missing values analysis
missing = df.isnull().sum()
missing = missing[missing > 0].sort_values(ascending=False)

if len(missing) > 0:
    print("\nFeatures with Missing Values (Top 10):")
    for col, count in missing.head(10).items():
        pct = (count / len(df)) * 100
        print(f"   {col}: {count:,} ({pct:.1f}%)")
    
    print("\nNote: Missing values expected for features requiring long lookback periods")
    print("(e.g., 60-day volatility needs 60 days of history)")
else:
    print("\nNo missing values detected after pipeline processing")

DATA QUALITY SUMMARY

Target Variable Distribution:
- Gold outperforms (target=1): 1,948 (42.8%)
- S&P outperforms (target=0): 2,600 (57.2%)
Class balance: Good (within 40-60% range)

Features with Missing Values (Top 10):
   oil_return_60d: 60 (1.3%)
   silver_return_60d: 60 (1.3%)
   sp500_vol_60d: 60 (1.3%)
   gold_vol_60d: 60 (1.3%)
   oil_vol_60d: 60 (1.3%)
   silver_vol_60d: 60 (1.3%)
   sp500_return_60d: 60 (1.3%)
   gold_return_60d: 60 (1.3%)
   cpi_yoy_change: 32 (0.7%)
   real_interest_rate: 32 (0.7%)

Note: Missing values expected for features requiring long lookback periods
(e.g., 60-day volatility needs 60 days of history)


## Step 4: Validate Price Ranges

Sanity check that prices fall within reasonable historical bounds (2006-2024).

In [13]:
# VALIDATION: Historical Price Ranges

print("Historical Price Ranges:")
print(f"- Gold:    ${df['gold_price'].min():.0f} - ${df['gold_price'].max():.0f}")
print(f"- S&P 500: {df['sp500_price'].min():.0f} - {df['sp500_price'].max():.0f}")
print(f"- Silver:  ${df['silver_price'].min():.2f} - ${df['silver_price'].max():.2f}")
print(f"\nAll prices within expected historical ranges (2006-2024)")

# Dataset coverage statistics
print(f"\nDataset Coverage:")
years_span = (df['date'].max() - df['date'].min()).days / 365.25
expected_trading_days = int(years_span * 252)  # ~252 trading days per year
coverage = (len(df) / expected_trading_days) * 100

print(f"Time span: {years_span:.1f} years")
print(f"Actual trading days: {len(df):,}")
print(f"Expected trading days: ~{expected_trading_days:,}")
print(f"Coverage: {coverage:.1f}% (accounts for weekends, holidays)")

Historical Price Ranges:
- Gold:    $528 - $2361
- S&P 500: 676 - 5254
- Silver:  $8.79 - $48.60

All prices within expected historical ranges (2006-2024)

Dataset Coverage:
Time span: 18.3 years
Actual trading days: 4,548
Expected trading days: ~4,603
Coverage: 98.8% (accounts for weekends, holidays)


## Step 5: Feature Category Breakdown

Organize features by type to understand the model's input space:
- **Raw prices:** Gold, S&P 500, silver, oil, USD, etc.
- **Returns:** 5-day, 20-day, 60-day momentum indicators
- **Moving averages:** 20-day trend indicators
- **Volatility:** 20-day and 60-day rolling volatility
- **Macro indicators:** CPI, unemployment, Fed Funds, M2, VIX, NASDAQ
- **Derived features:** Real interest rate, gold/silver ratio, yield curve slope

In [14]:
# FEATURE ANALYSIS: Categorize by Type

all_cols = df.columns.tolist()

# Define feature categories
price_cols = [c for c in all_cols if 'price' in c.lower() and 'return' not in c and 'vol' not in c and 'ma' not in c]
return_cols = [c for c in all_cols if 'return' in c]
ma_cols = [c for c in all_cols if '_ma_' in c]
vol_cols = [c for c in all_cols if '_vol_' in c and 'vol.' not in c]
macro_cols = [c for c in all_cols if any(x in c for x in ['cpi', 'unemployment', 'fedfunds', 'm2', 'oil', 'usd', 'vix', 'treasury', 'nasdaq'])]
derived_cols = [c for c in all_cols if c in ['yield_curve_slope', 'real_interest_rate', 'gold_silver_ratio', 'cpi_yoy_change']]

print("\nEngineered Feature Categories:")
print(f"\n   Raw Prices: {len(price_cols)} features")
print(f"      Example: {', '.join(price_cols[:3])}")

print(f"\n   Returns (momentum): {len(return_cols)} features")
print(f"      Windows: 5-day, 20-day, 60-day")

print(f"\n   Moving Averages (trend): {len(ma_cols)} features")
print(f"      Windows: 20-day")

print(f"\n   Volatility (risk): {len(vol_cols)} features")
print(f"      Annualized rolling standard deviation (20-day, 60-day)")

print(f"\n   Macro Indicators: {len(macro_cols)} features")
print(f"      Example: {', '.join([c for c in macro_cols if not c.endswith('_value')][:3])}")

print(f"\n   Derived Features: {len(derived_cols)} features")
print(f"      {', '.join(derived_cols)}")

print(f"\n   Total Features: {len(all_cols) - 2} (excluding date and target)")

# Display sample rows
print("\nFirst 5 Rows of Processed Data:")
display(df.head())


Engineered Feature Categories:

   Raw Prices: 4 features
      Example: gold_price, sp500_price, silver_price

   Returns (momentum): 12 features
      Windows: 5-day, 20-day, 60-day

   Moving Averages (trend): 4 features
      Windows: 20-day

   Volatility (risk): 8 features
      Annualized rolling standard deviation (20-day, 60-day)

   Macro Indicators: 16 features
      Example: treasury_yield, oil_price, oil_return_5d

   Derived Features: 4 features
      cpi_yoy_change, real_interest_rate, gold_silver_ratio, yield_curve_slope

   Total Features: 40 (excluding date and target)

First 5 Rows of Processed Data:


date  gold_price  sp500_price  silver_price  usd_index_value  \
0 2006-01-03       532.5       1268.8         9.160         100.7558   
1 2006-01-04       535.6       1273.5         9.170         100.2288   
2 2006-01-05       527.8       1273.5         8.872         100.2992   
3 2006-01-06       541.2       1285.5         9.173         100.0241   
4 2006-01-09       550.5       1290.2         9.283         100.1794   

   treasury_yield  nasdaq_value  vix_value  oil_price  gold_return_5d  ...  \
0            4.38       2243.74      11.14      61.63             NaN  ...   
1            4.35       2263.46      11.37      61.63             NaN  ...   
2            4.36       2276.87      11.31      61.63             NaN  ...   
3            4.38       2305.62      11.00      61.63             NaN  ...   
4            4.39       2318.69      11.13      61.63             NaN  ...   

   oil_vol_60d  cpi_cpiaucsl  unemployment_unrate  fedfunds_fedfunds  m2_m2sl  \
0          NaN           NaN                  NaN                NaN      NaN   
1          NaN           NaN                  NaN                NaN      NaN   
2          NaN           NaN                  NaN                NaN      NaN   
3          NaN           NaN                  NaN                NaN      NaN   
4          NaN           NaN                  NaN                NaN      NaN   

   cpi_yoy_change  real_interest_rate  gold_silver_ratio  yield_curve_slope  \
0             NaN                 NaN          58.133188               4.38   
1             NaN                 NaN          58.407852               4.35   
2             NaN                 NaN          59.490532               4.36   
3             NaN                 NaN          58.999237               4.38   
4             NaN                 NaN          59.301950               4.39   

   target  
0       1  
1       1  
2       1  
3       1  
4       1  

[5 rows x 42 columns]

## Processing Complete

### Summary

**Data Loading:** Loaded and merged 12 datasets (7 daily market + 5 monthly macro)  
**Feature Engineering:** Created 40+ features through returns, moving averages, volatility, and derived indicators  
**Mixed Frequency:** Properly handled daily and monthly data via forward-filling  
**Target Creation:** Generated binary classification target (gold outperforms S&P 500 over 90 days)  

---

### Data Quality Notes

- **Class Balance:** ~43% gold wins, ~57% stocks win
- **Missing Values:** Some features have missing values due to lookback windows (expected behavior)
- **Price Validation:** All prices validated against known historical bounds
- **Coverage:** ~98% of expected trading days captured (2006-2024)

---

### Next Steps

1. **02_eda.ipynb** - Exploratory data analysis and visualization
2. **03_modeling.ipynb** - Model training, evaluation, and validation

---

### Expected Modeling Outcome

Binary classifier predicting gold vs stock outperformance using time-based validation (2006-2016 train, 2017-2020 test).